CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/recipes/MCRecipeManager.java

```diff
@@ -301,132 +302,148 @@ public void removeRecipes(List<ResourceLocation> removingRecipes) {
-                RegistryManager.ACTIVE.getRegistry(GameData.RECIPES).remove(recipe);
-                recipesToAdd.stream()
-                        .filter(f -> f instanceof ActionDummyAddRecipe)
-                        .filter(f -> f.recipe.getRegistryName().equals(recipe))
-                        .forEach(toUnAdd::add);
-            });
-            toUnAdd.forEach(f -> {
-                recipesToAdd.remove(f);
-                usedRecipeNames.remove(f.getName());
-            });
+            if (ActionReplaceAllOccurences.INSTANCE.currentModifiedRecipe != null) {
+                if (forOutput == null || forOutput.contains(ActionReplaceAllOccurences.INSTANCE.currentModifiedRecipe.getOutput())) {
+                    if (ActionReplaceAllOccurences.INSTANCE.ingredients1D != null) {
+                        if (replaceWith == null)
+                            return; //No null's in shapeless recipes... We can't do anything, so we just won't modify the recipe.
+                        for (int i = 0; i < ActionReplaceAllOccurences.INSTANCE.ingredients1D.length; i++) {
+                            IIngredient ingredient = ActionReplaceAllOccurences.INSTANCE.ingredients1D[i];
+                            if (ingredient == null)
+                                continue;
+                            IIngredient changeResult = changeIngredient(ingredient);
+                            if (changeResult != ingredient) {
+                                ActionReplaceAllOccurences.INSTANCE.ingredients1D[i] = changeResult;
+                                ActionReplaceAllOccurences.INSTANCE.recipeModified = true;
+                            }
+                        }
+                    }
+                    if (ActionReplaceAllOccurences.INSTANCE.ingredients2D != null) {
+                        for (int i = 0; i < ActionReplaceAllOccurences.INSTANCE.ingredients2D.length; i++) {
+                            for (int j = 0; j < ActionReplaceAllOccurences.INSTANCE.ingredients2D[i].length; j++) {
+                                IIngredient ingredient = ActionReplaceAllOccurences.INSTANCE.ingredients2D[i][j];
+                                if (ingredient == null)
+                                    continue;
+                                IIngredient changeResult = changeIngredient(ingredient);
+                                if (changeResult != ingredient) {
+                                    ActionReplaceAllOccurences.INSTANCE.ingredients2D[i][j] = changeResult;
+                                    ActionReplaceAllOccurences.INSTANCE.recipeModified = true;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
         }
-        
+
         @Override
         public String describe() {
-            return "Removing all occurences of ingredient: " + toReplace + " and replacing them with " + replaceWith;
+            return null;
```

Comment:
Why does this return null?

Reply:
because I traversal all recipes and apply the action, every recipe requires apply the action once. I don't want the print the describe message 10 thousand times.